magento/pwa-studio
# [SPIKE] [JIRA-250] Sign out alert extension using Peregrine Tap Hooks.

## Description
This PR tries to approach the peregrine extensibility topic using a page from webpack's plugin system.

This framework is implemented using webpack's `tapable` package. The `tapable` package provides a sophisticated API to implement call handlers and interceptors.

**The scenario that is used as the basis for this SPIKE goes like this:**
`Assume there is an extension provider who would like to write an extension that when added to Venia, provides an alert to the user once they have clicked on the signout button.`

**Solution:**
We add a Sync hook to the `useMyAccount` which will also be provided to all extensions (in our case the `signoutExtension.js`) so they can tap and register a callback for it. When the user clicks on the sign out button, the sync hook will be called and all the register extensions will get a chance to handle it.

`Tapable` is a very complex looking package at first but it is very very useful. (Literally webpack runs on it  😐).

We can come up with a formalized way to handle and provide these tapable hooks if the team believes this is the right approach to handle peregrine extensibility.

Feel free to drop your thoughts on this PR. Cheers guys.

## Related Issue
Closes [JIRA-250](https://jira.corp.magento.com/browse/PWA-250)

### Verification Steps
1. Go to the deployed page and sign in.
2. Once signed in, click sign out.
3. You should see an alert that says `Signing Out`.

## Screenshots / Screen Captures (if appropriate)
![image](https://user-images.githubusercontent.com/35203638/71216217-8bced380-227f-11ea-9cda-fdd155ebd55d.png)

## Checklist
None
